src/GafferScene/Instancer.cpp

```diff
@@ -482,18 +483,63 @@ class Instancer::EngineData : public Data
 					for( size_t i = 0, e = numPoints(); i < e; ++i )
 					{
 						int id = (*m_ids)[i];
-						if( m_idsToPointIndices.try_emplace( id, i ).second )
+						auto ins = m_idsToPointIndices.try_emplace( id, i );
+						if( !ins.second )
 						{
-							if( constantPrototypeIndex != -1 )
+							if( !omitDuplicateIds )
 							{
-								instancesForPrototypeIndex[ constantPrototypeIndex ].push_back( i );
-								continue;
+								throw IECore::Exception( fmt::format( "Instance id \"{}\" is duplicated at index {} and {}. This probably indicates invalid source data, if you want to hack around it, you can set \"omitDuplicateIds\".", id, m_idsToPointIndices[id], i ) );
 							}
+							hasDuplicates = true;
+							ins.first->second = std::numeric_limits<size_t>::max();
+						}
+
+						if( hasDuplicates )
+						{
+							continue;
+						}
+
+						if( constantPrototypeIndex != -1 )
+						{
+							instancesForPrototypeIndex[ constantPrototypeIndex ].push_back( i );
+							continue;
+						}
+
+						int prototypeIndex = m_prototypeIndexRemap[ (*m_prototypeIndices)[ i ] % m_numPrototypes ];
+						if( prototypeIndex != -1 )
+						{
+							instancesForPrototypeIndex[ prototypeIndex ].push_back( i );
+						}
+					}
 
-							int prototypeIndex = m_prototypeIndexRemap[ (*m_prototypeIndices)[ i ] % m_numPrototypes ];
-							if( prototypeIndex != -1 )
+					if( hasDuplicates )
```

Comment:
I have to be honest and say that I haven't really wrapped my head round this - it feels kindof convoluted. Why mix up the code for building `m_idsToPointIndices` with the code for building `instancesForPrototypeIndex` (see earlier comment) if we then have to back out and have a special case that computes the latter from the former afterwards anyway? There may well be a good reason, but as a slightly fuzzy headed reader I'm finding it hard to grasp.

Reply:
As mentioned in reply to other comments, I did manage to refactor this in a way I suspect you will be happier with.